### ISSUE TITLE
Can't yield a block in initialize

### ISSUE THREAD

**Issue by fazibear** [opal/opal#1176]
I think this a bug, pass a block to initialize but can't yield. `block_given?` is also `false`

```ruby
class A
  def initialize(opts)
    yield "OK"
  end
end

x = Native(`{a: {b: "asd"}}`)

A.new(x.a.b) do |s|
  puts s
end
```

Throws exception `Uncaught LocalJumpError: no block given`

And little work around:

```ruby
class A
  def initialize(opts)
    yield "OK"
  end
end

x = Native(`{a: {b: "asd"}}`)

c = x.a.b

A.new(c) do |s|
  puts s
end
```

**Comment by meh** [comment]
This sounds like #895.

**Comment by meh** [comment]
@vais /cc

**Comment by elia** [comment]
Maybe it's something similar to https://github.com/opal/opal/commit/7cf471534c2a4b826d06a328008a9c8fd393493f

**Comment by vais** [comment]
@meh this one is truly bizarre, need your help as you are the world's foremost expert on the `Native` module :wink:

This code:
```ruby
require 'native'

class A
  def initialize(opts)
    yield "OK"
  end
end

x = Native(`{y: {z: "asd"}}`)

A.new(x.y.z) do |s|
  puts s
end
```
compiles to this JS:
```javascript
/* Generated by Opal 0.9.0.beta1 */
(function(Opal) {
  Opal.dynamic_require_severity = "error";
  var OPAL_CONFIG = { method_missing: true, arity_check: false, freezing: true, tainting: true };
  var $a, $b, TMP_2, self = Opal.top, $scope = Opal, nil = Opal.nil, $breaker = Opal.breaker, $slice = Opal.slice, $klass = Opal.klass, x = nil;

  Opal.add_stubs(['$require', '$Native', '$new', '$puts', '$z', '$y']);
  self.$require("native");
  (function($base, $super) {
    function $A(){};
    var self = $A = $klass($base, $super, 'A', $A);

    var def = self.$$proto, $scope = self.$$scope, TMP_1;

    return (Opal.defn(self, '$initialize', TMP_1 = function(opts) {
      var $a, self = this, $iter = TMP_1.$$p, $yield = $iter || nil;

      TMP_1.$$p = null;
      return $a = Opal.yield1($yield, "OK"), $a === $breaker ? $a : $a;
    }), nil) && 'initialize'
  })($scope.base, null);
  x = self.$Native({y: {z: "asd"}});
  return ($a = ($b = $scope.get('A')).$new, $a.$$p = (TMP_2 = function(s){var self = TMP_2.$$s || this;
if (s == null) s = nil;
  return self.$puts(s)}, TMP_2.$$s = self, TMP_2), $a).call($b, x.$y().$z());
})(Opal);
```
And fails with:
```
/vagrant/opal/vais.js:999
      throw Opal.LocalJumpError.$new("no block given");
                                ^
LocalJumpError: no block given
    at OpalClass.Opal.defs.self [as $new] (/vagrant/opal/vais.js:4432:15)
    at Opal.yield1 (/vagrant/opal/vais.js:999:33)
    at $A.Opal.defn.TMP_1 [as $initialize] (/vagrant/opal/vais.js:19767:24)
    at OpalClass.Opal.defn.TMP_2 [as $new] (/vagrant/opal/vais.js:2793:23)
    at Opal.dynamic_require_severity (/vagrant/opal/vais.js:19773:56)
    at Object.<anonymous> (/vagrant/opal/vais.js:19774:3)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
```
After instrumenting the compiled JS a little as follows:
```javascript
...
if (s == null) s = nil;
  return self.$puts(s)}, TMP_2.$$s = self, TMP_2), console.log(">>> Let's see $a.$$p:\n\n", $a.$$p, "\n\n>>> So far so good, now let's call the 'native' method x.$y():\n\n", x.$y(), "\n\n>>> Ok, now let's see about $a.$$p again:\n\n", $a.$$p, "\n\n>>> WTF?! Where'd it go??"), $a).call($b);
})(Opal);
...
```
We get this:
```
>>> Let's see $a.$$p:

 function (s){var self = TMP_2.$$s || this;
if (s == null) s = nil;
  return self.$puts(s)}

>>> So far so good, now let's call the 'native' method x.$y():

 { '$$id': 276, native: { z: 'asd' } }

>>> Ok, now let's see about $a.$$p again:

 null

>>> WTF?! Where'd it go??

/vagrant/opal/vais.js:999
      throw Opal.LocalJumpError.$new("no block given");
                                ^
LocalJumpError: no block given
    at OpalClass.Opal.defs.self [as $new] (/vagrant/opal/vais.js:4432:15)
    at Opal.yield1 (/vagrant/opal/vais.js:999:33)
    at $A.Opal.defn.TMP_1 [as $initialize] (/vagrant/opal/vais.js:19767:24)
    at OpalClass.Opal.defn.TMP_2 [as $new] (/vagrant/opal/vais.js:2793:23)
    at Opal.dynamic_require_severity (/vagrant/opal/vais.js:19773:277)
    at Object.<anonymous> (/vagrant/opal/vais.js:19774:3)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
```
(Notice that I did not have to do `.call($b, x.$y().$z())`, just `.call($b)`, and it still fails because `x.$y()` is getting executed inside `console.log()`. The point at which `x.$y()` is executed is important. In fact, if we stash the value of `$a.$$p` prior to the call to `x.$y()` and then restore it immediately afterwards, there is no failure.)

So, basically, what's happening, is that **somehow** the act of calling `x.$y()` is robbing `$a` of its `$$p` property. Now, how in the world is that even possible? One answer would be a leaky global, but there are none here. By what other mechanism in javascript could a method call on object A remove a property of object B to which object A has no reference? *That* is beyond me.

**Comment by elia** [comment]
Seems like the `$$p` prop is cleared somewhere…

I'm under the impression that finding a way to move the block passing to the last argument would not only make the code much more readable, but also solve this kind of bugs by not using `.$$p` which being shared between calls is subject to race-condition problems.

_Anyway it could be of some help to remind everyone that `opal -R server issue-1176.rb` will serve on port 3000 the passed code allowing to use the browser debugger to navigate code._

**Comment by vais** [comment]
@elia but the `$$p` is a property on an object that is not shared with the other code, so how...? I don't even... :confused: 

Re `opal -R server issue-1176.rb` - I had no idea, this is wonderful!

**Comment by elia** [comment]
But methods are shared, `$new` in our case is `Opal.Class.$$proto.$new` even when you call it as `Opal.A.$new`

**Comment by vais** [comment]
Oh, crap, right...

**Comment by vais** [comment]
Maybe in the meantime @meh can look into what's causing the reset in `Native`. The change you're proposing is quite drastic I imagine...

**Comment by elia** [comment]
indeed, it's something we talked about quite a bit in the past, surely not something that should ship in 0.9

**Comment by vais** [comment]
I just hope @meh is feeling bored today :stuck_out_tongue_winking_eye:
